Adds Web3.py page

albertov19 opened the pull request:
### Description

Please explain the changes this PR addresses here.

### Checklist

- [x] If this page requires a disclaimer, I have added one
- [ ] If pages have been moved around, I have created an additional PR in `moonbeam-mkdocs` to update redirects

eshaben commented on .snippets/code/ethereum-api/web3py/balances.py:
```diff
@@ -0,0 +1,20 @@
+web3 = Web3(Web3.HTTPProvider(provider_rpc['evm_containerchain']))
+
+# 2. Create address variables
+address_from = 'ADDRESS_FROM_HERE'
```
Made this comment on the Ethers PR, but variable placeholders should start with `INSERT` following by whatever you're using to describe the variable and they should not end with `HERE`

so this should be `INSERT_FROM_ADDRESS` (made this change on the moonbeam-docs repo already)

eshaben commented on .snippets/text/common/general-py-tutorial-check.md:
```diff
@@ -0,0 +1,5 @@
+!!! note
+    The examples in this guide are based on a MacOS or Ubuntu 20.04 environment. If you're using Windows, you'll need to adapt them accordingly.
+    
+    Furthermore, please ensure that you have Python3 and a package manager like PIP installed.
```
I don't think PIP is ever capitalized like this 🤔 probably not a big deal, but it just looked a little off

eshaben commented on builders/interact/ethereum-api/libraries/web3py.md:
```diff
@@ -0,0 +1,264 @@
+---
+title: Send Transactions & Deploy Contracts with Web3.py
```
This ends up being "Send Transactions & Deploy Contracts with Web3.py | Tannsi Docs", which makes the title too long. It needs to be 60 characters or less

albertov19 replied:
I did `EVM Transactions & Contracts with Web3.py`

eshaben commented on builders/interact/ethereum-api/libraries/web3py.md:
```diff
@@ -0,0 +1,264 @@
+---
+title: Send Transactions & Deploy Contracts with Web3.py
+description: Follow this tutorial to learn how to use the Ethereum Web3 Python Library to send transactions and deploy Solidity smart contracts to your Tanssi EVM ContainerChain.
```
Descriptions should be 160 characters or less

eshaben commented on builders/interact/ethereum-api/libraries/web3py.md:
````diff
@@ -0,0 +1,264 @@
+
+For this guide, you'll need to install the Web3.py library and the Solidity compiler. To install both packages, you can run the following command:
+
+```
````
languages should be applied to all code snippets wrapped in triple back ticks. This should be `bash`

eshaben reviewed (commented):
Just left a few general comments on some standards. Didn't review very closely so don't rely entirely on this review haha

Side note, we need to be careful of modifying the colors of brand logos. Some brands have very strict guidelines on what can and cannot be modified in terms of their logo (i.e., Moonbeam hahaha)

augustocollado commented on builders/interact/ethereum-api/libraries/web3py.md:
```diff
@@ -0,0 +1,264 @@
+2. [Set up the Web3 provider](#setting-up-the-web3py-provider)
+3. Define the `account_from`, including the `private_key`, the `contract_address` of the deployed contract, and the `value` to increment by. The private key is required to sign the transaction. **Note: This is for example purposes only. Never store your private keys in a Python file**
+4. Create a contract instance using the `web3.eth.contract` function and passing in the ABI and address of the deployed contract
+5. Build the increment transaction using the contract instance and passing in the value to increment by. You'll then use the `build_transaction` function to pass in the transaction information including the `from` address and the `nonce` for the sender. To get the `nonce` you can use the `web3.eth.get_transaction_count` function
```
This step fails (at least on my computer) when calling the method "web3.eth.get_transaction_count", with exception web3.exceptions.InvalidAddress: ('web3.py only accepts checksum addresses blah...')
To solve this, converting the address to its checksum version works:
web3.eth.get_transaction_count( Web3.to_checksum_address(account_from['address']) )

augustocollado replied:
Looks like the 'from' parameter from the tx also requires a checksum address. the complete script could be changed to something like:
# 5. Build increment tx
checksum_address = Web3.to_checksum_address(account_from['address'])
increment_tx = Incrementer.functions.increment(value).build_transaction(
    {
        'from': checksum_address,
        'nonce': web3.eth.get_transaction_count(checksum_address),
    }
)

albertov19 replied:
I don't hit this issue but I can add it so that it is compatible with more versions/OS. Updated both snippets for Increment and Reset files

augustocollado commented on builders/interact/ethereum-api/libraries/web3py.md:
```diff
@@ -0,0 +1,264 @@
+2. [Set up the Web3 provider](#setting-up-the-web3py-provider)
+3. Define the `account_from`, including the `private_key`, and the `contract_address` of the deployed contract. The private key is required to sign the transaction. **Note: This is for example purposes only. Never store your private keys in a Python file**
+4. Create a contract instance using the `web3.eth.contract` function and passing in the ABI and address of the deployed contract
+5. Build the reset transaction using the contract instance. You'll then use the `build_transaction` function to pass in the transaction information including the `from` address and the `nonce` for the sender. To get the `nonce` you can use the `web3.eth.get_transaction_count` function
```
Same problem as the previous script (at least on my computer). Solved it changing the code to:

checksum_address = Web3.to_checksum_address(account_from['address'])
reset_tx = Incrementer.functions.reset().build_transaction(
    {
        'from': checksum_address,
        'nonce': web3.eth.get_transaction_count(checksum_address),
    }
)

augustocollado reviewed (changes requested):
Brilliant, clear and easy to follow. 
I left some comments about an exception I got while following the steps (along with the fix)

albertov19 commented on builders/interact/ethereum-api/libraries/web3py.md:
```diff
@@ -0,0 +1,264 @@
+
+1. Add imports, including Web3.py and the ABI of the `Incrementer.sol` contract
+2. [Set up the Web3 provider](#setting-up-the-web3py-provider)
+3. Define the `account_from`, including the `private_key`. The private key is required to sign the transaction. **Note: This is for example purposes only. Never store your private keys in a Python file**
```
Good catch!

albertov19 commented:
@eshaben @augustocollado I implemented Black Python Formatter for Python files as it was not loading anything I would throw to `.editorconfig` - I added a note in the `.editorconfig` to ensure it is clear

augustocollado reviewed (approved):
lgtm!
